implementation_details.md

```diff
@@ -0,0 +1,160 @@
+
+# Implementation Details
+## `Connection` != `Player`
+I know I've been using the terms "client" and "player" somewhat interchangeably, but `Connection` and `Player` should be separate tokens. There's no benefit in forcing one player per connection. Having `Player` be its own thing makes it easier to do stuff like online splitscreen, temporarily fill team slots with bots, etc.
+
+## "Clock" Synchronization
+Ideally, clients predict ahead by just enough to have their inputs reach the server right before they're needed. People often try to have clients estimate the clock time on the server (with some SNTP handshake) and use that to schedule the next simulation step, but that's overly complex.
+
+What we really care about is: How much time passes between when the server receives my input and when that input is consumed? If the server simply tells clients how long their inputs are waiting in its buffer, the clients can use that information to converge on the correct lead.
```

Comment:
Can clients just estimate by how much they need to be ahead by calculating this value based on RTT? They can usually guess the current frame number on server (assuming it comes with updates), they know how much time on average it takes for their packets to be acknowledged and they can also estimate jitter and packet loss.

Or is it a good practice for servers to tell by how much a client has to be ahead explicitly?

Reply:
IMO guessing is more work, not less.

What we're trying to control is the predict-ahead amount. It's simpler and more accurate to just do that directly.

RTT can only be approximated whereas the server can measure the exact duration inputs spend in its buffer (more informative too since it "contains" both latency and jitter). Loss can be interpreted as a negative duration. Loss doesn't need to be estimated either.

> Or is it a good practice for servers to tell by how much a client has to be ahead explicitly?

This can be an entirely client-sided process. In my example, the server only tells clients how much they *are* ahead (sent with the updates). 
Clients could decide for themselves how much *to be* ahead.